### PR TITLE
Cancel task edit resets the task to the original values without refreshing

### PR DESCRIPTION
**Pull request by Hus274**
#7688
### Changes

Updating the cancel functionality of editing a task in order to reset the task to the original values if the user cancels the edit.

---

UUID: 2a4d30ee-4833-420e-be7c-ad6a6e261c5f

### CODE REVIEW

File: test/spec/controllers/tasksCtrlSpec.js [hunk 1]
```diff
@@ -30,11 +30,89 @@ describe('Tasks Controller', function() {
...
+
+    it('sets _tags to false if preference for collapsed tags is turned on', function() {
+      user.preferences.tagsCollapsed = true;
+      scope.editTask(task, user);
+
+      expect(task._tags).to.eql(false);
+    });
+
+    it('sets _advanced to true by default', function(){
+      user.preferences.advancedCollapsed = true;
```
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: You should update the `specHelper.newUser()` function to include this default setting.  [Here](https://github.com/HabitRPG/habitrpg/pull/6943/files#diff-ecd2e89a7b19d9ac70c6d589bb181bd9R39) is an example of where this has been done before.

File: test/spec/controllers/tasksCtrlSpec.js [hunk 2]
```diff
@@ -30,11 +30,89 @@ describe('Tasks Controller', function() {
...
+      scope.editTask(task, user);
+
+      expect(task._advanced).to.eql(false);
```
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: Test description says `true` but you assert `false`?

Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: I think your test descriptions for this test and the next one are mixed up.

File: test/spec/controllers/tasksCtrlSpec.js (hunk 1 above)
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: Whatever the default setting should actually be, not necessarily this setting. :stuck_out_tongue_winking_eye: (See below comments.)

File: test/spec/controllers/tasksCtrlSpec.js [hunk 3]
```diff
@@ -30,11 +30,89 @@ describe('Tasks Controller', function() {
...
+    });
+
+    it('sets _advanced to false if preference for collapsed advance menu is turned on', function() {
+      user.preferences.advancedCollapsed = false;
+      scope.editTask(task, user);
+
+      expect(task._advanced).to.eql(true);
     });
+
```
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: Extra space.

File: test/spec/controllers/tasksCtrlSpec.js [hunk 4]
```diff
@@ -30,11 +30,89 @@ describe('Tasks Controller', function() {
...
+    var task;
+
+    beforeEach(function(){
+      task = specHelper.newTask();
+    });
+
+    it('resets the task text', function() {
+      let originalText = task.text;
+      scope.editTask(task, user);
+      task.text = 'test';
```
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: Sanity check: assert that `originalText` is not `'test'`.

File: test/spec/controllers/tasksCtrlSpec.js [hunk 5]
```diff
@@ -30,11 +30,89 @@ describe('Tasks Controller', function() {
...
+      scope.cancelTaskEdit(task)
```
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: missing semicolon

File: test/spec/controllers/tasksCtrlSpec.js [hunk 6]
```diff
@@ -30,11 +30,89 @@ describe('Tasks Controller', function() {
...
+      expect(task.text).to.be.eql(originalText);
+    });
+
+    it('resets the task notes', function() {
+      let originalNotes = task.notes;
+      scope.editTask(task, user);
+      task.notes = 'test';
```
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: Same sanity check as above

File: test/spec/controllers/tasksCtrlSpec.js [hunk 7]
```diff
@@ -30,11 +30,89 @@ describe('Tasks Controller', function() {
...
+      scope.cancelTaskEdit(task)
+      expect(task.notes).to.be.eql(originalNotes);
+    });
+
+    it('resets the task alias', function() {
+      task.alias = 'alias';
+      let originalAlias = task.alias;
+      scope.editTask(task, user);
+      task.alias = 'test';
```
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: Same sanity check as above

File: website/client/js/controllers/tasksCtrl.js [hunk 8]
```diff
@@ -100,8 +99,17 @@ habitrpg.controller("TasksCtrl", ['$scope', '$rootScope', '$location', 'User','N
       }
     };
 
+    $scope.editTask = function(task, user) {
+      originalTasks.push(_.cloneDeep(task));
+      task._editing = !task._editing;
+      task._tags = !user.preferences.tagsCollapsed;
```
Comment (khipkin) [inline website/client/js/controllers/tasksCtrl.js]: Make sure `specHelper.newUser()` sets default values for both of these settings.

File: website/client/js/controllers/tasksCtrl.js [hunk 9]
```diff
@@ -100,8 +99,17 @@ habitrpg.controller("TasksCtrl", ['$scope', '$rootScope', '$location', 'User','N
...
+      task._advanced = !user.preferences.advancedCollapsed;
+      if($rootScope.charts[task._id]) $rootScope.charts[task.id] = false;
```
Comment (khipkin) [inline website/client/js/controllers/tasksCtrl.js]: Is this check necessary?

File: test/spec/controllers/tasksCtrlSpec.js (hunk 4 above)
Comment (crookedneighbor) [inline test/spec/controllers/tasksCtrlSpec.js]: That sanity check doesn't make sense. originalText would be a string, so it doesn't hold onto a reference like an object would.

File: website/client/js/services/taskServices.js [hunk 10]
```diff
@@ -150,13 +150,6 @@ angular.module('habitrpg')
       });
     };
 
-    function editTask(task, user) {
```
Comment (crookedneighbor) [inline website/client/js/services/taskServices.js:153]: Why are you removing this from the task service? The challenges controller also uses this function.

File: website/client/js/controllers/tasksCtrl.js [hunk 11]
```diff
@@ -100,8 +99,17 @@ habitrpg.controller("TasksCtrl", ['$scope', '$rootScope', '$location', 'User','N
       }
     };
 
+    $scope.editTask = function(task, user) {
+      originalTasks.push(_.cloneDeep(task));
```
Comment (crookedneighbor) [inline website/client/js/controllers/tasksCtrl.js]: This seems like a weird way to deal with this. I'd model it after how groups are edited.

File: test/spec/controllers/tasksCtrlSpec.js (hunk 4 above)
Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: But the originalText string might also be 'test'?

Comment (crookedneighbor) [inline test/spec/controllers/tasksCtrlSpec.js]: Hmm... that's true, from the spec helper.

Comment (khipkin) [inline test/spec/controllers/tasksCtrlSpec.js]: Exactly!  The sanity check is to make sure you aren't "changing" a string to the exact same string.

**Comment by crookedneighbor** [comment]
Basically, I'd use the angular copy function to copy the task and save it to `task._edit`. Then change the edit views to reference `task._edit.whatever` instead of the properties themselves. 

When editing is cancelled, you just close the edit lane as normal and the task shows as it should. 

If it's saved, update the task with the response back from the server (might already be don't that)

**Comment by Hus274** [comment]
Ahh that's a great idea, and it eliminates the possibility of the memory leak which I didn't love. I can work on getting that up and cleaning up based on the other comments and then I'll repost. Thanks for the assistance!

**Comment by Hus274** [comment]
Updated according to Blade's comments. The tests are failing but they are failing for me on the main branch as well so I'm not quite sure what's wrong.

**Comment by Hus274** [comment]
Looks like the test is failing because it's missing the apn npm module. After installing this module locally, the tests all passed.

File: test/spec/services/taskServicesSpec.js [hunk 12]
```diff
@@ -161,8 +161,6 @@ describe('Tasks Service', function() {
     it('toggles the _editing property', function() {
```
Comment (crookedneighbor) [inline test/spec/services/taskServicesSpec.js]: This is no longer accurate. You made it so that it doesn't toggle it. It just sets it to true.

**Comment by Alys** [comment]
@Hus274 There's a feature where a task being edited stays visible while it's open for editing even if it wouldn't normally be visible otherwise. E.g., if it's tagged "morning" and you have the "morning" tag filter active, and you edit the task to deselect that tag, the task does not disappear from view until after you've saved the edit. Another example, if there's a Daily active/grey filter active, and you change the due status of the Daily, the Daily stays visible while editing is happening. Can you test that that still works? I haven't seen anything in your code to indicate that it won't, but a quick test will be more reliable than me reading the code. :)

**Comment by Hus274** [comment]
@Alys I checked both these cases, the dailies filter and the tags filter, and they both worked as expected! Definitely worth checking it out and doing the test, thanks for bringing it up! :)

**Comment by Alys** [comment]
@Hus274 Awesome, thanks for testing! Trying to edit tasks was a bit of a nightmare before that feature was put in place so it's good to know it's still there. :)

File: test/spec/services/taskServicesSpec.js [hunk 13]
```diff
@@ -150,6 +150,56 @@ describe('Tasks Service', function() {
...
+
+    it('sets _tags to false if preference for collapsed tags is turned on', function() {
+      user.preferences.tagsCollapsed = true;
+      tasks.editTask(task, user);
+
+      expect(task._tags).to.eql(false);
+    });
+
+    it('sets _advanced to true by default', function(){
+      user.preferences.advancedCollapsed = true;
```
Comment (khipkin) [inline test/spec/services/taskServicesSpec.js:180]: add default preference value to `SpecHelper.newUser()`

File: website/client/js/controllers/tasksCtrl.js [hunk 14]
```diff
@@ -196,16 +191,15 @@ habitrpg.controller("TasksCtrl", ['$scope', '$rootScope', '$location', 'User','N
...
 
     $scope.addChecklistItem = function(task, $event, $index) {
-      if (task.checklist[$index].text) {
-        $scope.saveTask(task, true);
-        if ($index === task.checklist.length - 1)
-          task.checklist.push({ completed: false, text: '' });
-        focusChecklist(task, $index + 1);
+      if (task._edit.checklist[$index].text) {
+        if ($index === task._edit.checklist.length - 1)
+          task._edit.checklist.push({ completed: false, text: '' });
```
Comment (khipkin) [inline website/client/js/controllers/tasksCtrl.js:201]: add curly braces

File: test/spec/services/taskServicesSpec.js (hunk 13 above)
Comment (crookedneighbor) [inline test/spec/services/taskServicesSpec.js:180]: Not necessary.

File: test/spec/services/taskServicesSpec.js [hunk 15]
```diff
@@ -200,6 +198,21 @@ describe('Tasks Service', function() {
     });
   });
 
+  describe('cancelTaskEdit', function() {
+
```
Comment (crookedneighbor) [inline test/spec/services/taskServicesSpec.js]: Remove this extra line break.

**Comment by Hus274** [comment]
Updated formatting according to comments.

**Comment by crookedneighbor** [comment]
Notated toward your next tier.

**Pull request merged**